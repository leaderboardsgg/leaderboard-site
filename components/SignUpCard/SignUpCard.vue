<template>
  <BaseCard
    id="signUpCard"
    data-testid="sign-up-card"
    class="bg-white w-full max-w-xl"
  >
    <BaseCardHeader class="flex flex-row space-x-3">
      <BaseButton
        id="login-button"
        class="
          flex flex-1
          justify-center
          bg-white
          text-gray-900
          border border-gray-300
        "
        @click="$emit('logInClick')"
      >
        Log In
      </BaseButton>

      <div
        class="
          flex flex-1
          justify-center
          px-3
          py-2
          rounded
          bg-gray-100
          text-gray-900
        "
      >
        Sign Up
      </div>

      <CoreCloseButton v-show="modal" @click.prevent="$emit('close')" />
    </BaseCardHeader>

    <!-- Card Body -->
    <div class="p-3">
      <div class="flex flex-col space-y-3 pb-3 mb-3 border-b border-gray-300">
        <BaseInput
          v-model="email"
          name="email"
          type="text"
          placeholder="Email"
          autocomplete="email"
        />

        <div class="flex flex-1 flex-col space-y-1">
          <BaseInput
            v-model="username"
            name="username"
            type="text"
            placeholder="Username"
            autocomplete="nickname"
          />
          <p class="text-sm">This can be changed later</p>
        </div>

        <div class="flex flex-1 flex-col space-y-1">
          <div
            class="
              flex flex-1 flex-col
              sm:flex-row
              space-y-3
              sm:space-y-0 sm:space-x-2
            "
          >
            <BaseInput
              v-model="password"
              name="password"
              class="flex-grow"
              :type="showPassword ? 'text' : 'password'"
              placeholder="Password"
              autocomplete="password"
            />

            <BaseInput
              v-model="confirmPassword"
              name="confirmPassword"
              class="flex-grow"
              :type="showPassword ? 'text' : 'password'"
              placeholder="Confirm"
              autocomplete="password"
            />

            <HideShowPassword
              type="button"
              @click="showPassword = !showPassword"
              @keydown.enter.prevent
              @keyup.enter="showPassword = !showPassword"
            />
          </div>

          <p class="text-sm">
            * Must be 8 characters with a mix of letters and numbers
          </p>
        </div>

        <BaseButton class="signup-button"> Sign Up </BaseButton>
      </div>

      <div class="flex flex-col w-full space-y-2">
        <BaseButton class="signup-button">
          <svg
            class="w-5 h-5 mr-2"
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 480 512"
          >
            <path
              fill="currentColor"
              d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1zM480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2zm-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3zm-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1z"
            ></path>
          </svg>

          <p>Sign Up with Github</p>
        </BaseButton>

        <BaseButton class="signup-button">
          <svg
            class="w-5 h-5 mr-2"
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 488 512"
          >
            <path
              fill="currentColor"
              d="M488 261.8C488 403.3 391.1 504 248 504 110.8 504 0 393.2 0 256S110.8 8 248 8c66.8 0 123 24.5 166.3 64.9l-67.5 64.9C258.5 52.6 94.3 116.6 94.3 256c0 86.5 69.1 156.6 153.7 156.6 98.2 0 135-70.4 140.8-106.9H248v-85.3h236.1c2.3 12.7 3.9 24.9 3.9 41.4z"
            ></path>
          </svg>

          <p>Sign Up with Google</p>
        </BaseButton>
      </div>
    </div>
  </BaseCard>
</template>

<script lang="ts">
import { defineComponent, reactive, toRefs } from '@nuxtjs/composition-api';
import HideShowPassword from '../HideShowPassword/HideShowPassword.vue';

export default defineComponent({
  components: {
    HideShowPassword,
  },
  props: {
    modal: {
      type: Boolean,
      required: false,
      default: false,
    },
  },

  setup() {
    const state = reactive({
      confirmPassword: '',
      email: '',
      password: '',
      showPassword: false,
      username: '',
    });

    return {
      ...toRefs(state),
    };
  },
});
</script>

<style scoped>
#hide-show-password {
  @apply text-gray-700 hover:bg-gray-100;
}

#login-button {
  @apply hover:bg-gray-100;
}

.signup-button {
  @apply flex flex-1 items-center justify-center fill-current bg-gray-100 text-gray-900 hover:bg-gray-200;
}
</style>
