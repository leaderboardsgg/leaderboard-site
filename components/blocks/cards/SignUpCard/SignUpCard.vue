<script setup lang="ts">
import type { Ref } from 'vue'
import BaseInput from 'elements/inputs/BaseInput/BaseInput.vue'
import HideShowPassword from 'elements/buttons/HideShowPassword/HideShowPassword.vue'
import BaseButton from 'elements/buttons/BaseButton/BaseButton.vue'
import CloseButton from 'elements/buttons/CloseButton/CloseButton.vue'
import Card from 'elements/cards/Card/Card.vue'
import CardHeader from 'elements/cards/CardHeader/CardHeader.vue'
import CardBody from 'elements/cards/CardBody/CardBody.vue'
import { useRegisterUser } from 'composables/api'

interface SignUpCardProps {
  modal?: boolean
}

interface SignUpCardState {
  showPassword: Ref<boolean>
}

interface UserRegister {
  email: Ref<string>
  password: Ref<string>
  passwordConfirm: Ref<string>
  username: Ref<string>
}

const emit = defineEmits<{
  (event: 'close'): void
  (event: 'logInClick'): void
  (event: 'signUpClick'): void
}>()

const props = withDefaults(defineProps<SignUpCardProps>(), {
  modal: false,
})

const register: UserRegister = {
  email: ref(''),
  password: ref(''),
  passwordConfirm: ref(''),
  username: ref(''),
}

const state: SignUpCardState = {
  showPassword: ref(false),
}

function signup() {
  useRegisterUser({
    email: register.email.value,
    password: register.password.value,
    passwordConfirm: register.passwordConfirm.value,
    username: register.username.value,
  })

  register.email.value = ''
  register.password.value = ''
  register.passwordConfirm.value = ''
  register.username.value = ''
  state.showPassword.value = false

  emit('signUpClick')
}
</script>

<template>
  <Card id="signUpCard" data-testid="sign-up-card" class="signup-card">
    <CardHeader class="signup-card__header">
      <BaseButton
        id="login-button"
        class="signup-card__login-button"
        data-testid="login-button"
        @click="emit('logInClick')"
      >
        Log In
      </BaseButton>

      <div class="signup-card__title">Sign Up</div>

      <CloseButton
        v-show="props.modal"
        data-testid="close-button"
        @click.prevent="emit('close')"
      />
    </CardHeader>

    <CardBody>
      <div class="signup-card__body-wrapper">
        <BaseInput
          :model="register.email"
          name="email"
          type="text"
          placeholder="Email"
          autocomplete="email"
          data-testid="email-input"
        />

        <div class="signup-card__input-wrapper">
          <BaseInput
            :model="register.username"
            name="username"
            type="text"
            placeholder="Username"
            autocomplete="nickname"
            data-testid="username-input"
          />
          <p>This can be changed later</p>
        </div>

        <div class="signup-card__input-wrapper">
          <div class="signup-card__password-wrapper">
            <BaseInput
              :model="register.password"
              name="password"
              class="signup-card__password-field"
              :type="state.showPassword.value ? 'text' : 'password'"
              placeholder="Password"
              autocomplete="password"
              data-testid="password-input"
            />

            <BaseInput
              :model="register.passwordConfirm"
              name="passwordConfirm"
              class="signup-card__password-field"
              :type="state.showPassword.value ? 'text' : 'password'"
              placeholder="Confirm"
              autocomplete="password"
              data-testid="password-confirm-input"
            />

            <HideShowPassword
              id="hide-show-password"
              type="button"
              data-testid="hide-show-button"
              @click="state.showPassword.value = !state.showPassword.value"
              @keydown.enter.prevent=""
              @keyup.enter="
                state.showPassword.value = !state.showPassword.value
              "
            />
          </div>

          <p>* Must be 8 characters with a mix of letters and numbers</p>
        </div>

        <BaseButton
          class="signup-button"
          data-testid="sign-up-button"
          @click="signup"
        >
          Sign Up
        </BaseButton>
      </div>

      <div class="signup-card__auth-buttons">
        <BaseButton class="signup-button">
          <i-svg-github class="mr-2 h-5 w-5" />

          <p>Sign Up with Github</p>
        </BaseButton>

        <BaseButton class="signup-button">
          <i-svg-google class="mr-2 h-5 w-5" />

          <p>Sign Up with Google</p>
        </BaseButton>
      </div>
    </CardBody>
  </Card>
</template>

<style lang="postcss" scoped>
.signup-card {
  @apply bg-white w-full max-w-xl;

  & .signup-card__header {
    @apply flex flex-row space-x-3;
  }

  & .signup-card__title {
    @apply flex flex-1 justify-center px-3 py-2 rounded bg-gray-100 text-gray-900;
  }

  & .signup-card__login-button {
    @apply flex flex-1 justify-center bg-white text-gray-900 border border-gray-300;
  }

  & .signup-card__body-wrapper {
    @apply flex flex-col space-y-3 pb-3 mb-3 border-b border-gray-300;
  }

  & .signup-card__input-wrapper {
    @apply flex flex-1 flex-col space-y-1;
  }

  & .signup-card__password-wrapper {
    @apply flex flex-1 flex-col sm:flex-row space-y-3 sm:space-y-0 sm:space-x-2;
  }

  & .signup-card__password-field {
    @apply grow sm:w-full md:w-9/12;
  }

  & .signup-card__auth-buttons {
    @apply flex flex-col w-full space-y-2;
  }

  & p {
    @apply text-sm;
  }
}

#hide-show-password {
  @apply text-gray-700 hover:bg-gray-100;
}

#login-button {
  @apply hover:bg-gray-100;
}

.signup-button {
  @apply flex flex-1 items-center justify-center fill-current bg-gray-100 text-gray-900 hover:bg-gray-200;
}
</style>
