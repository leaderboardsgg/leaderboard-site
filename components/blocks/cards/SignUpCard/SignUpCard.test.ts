import { describe, expect, test } from 'vitest'
import { mount } from '@vue/test-utils'

import SignUpCard from './SignUpCard.vue'

describe('<SignUpCard />', () => {
  const SignUpCardWrapper = mount(SignUpCard)

  test('should render without crashing', () => {
    expect(SignUpCardWrapper.isVisible()).toBe(true)
  })

  // describe('when the close button is clicked', () => {
  //   test('should close the SignUpCard', async () => {
  //     await SignUpCardWrapper.get('[data-testid="close-button"]').trigger(
  //       'click',
  //     )

  //     expect(SignUpCardWrapper.isVisible()).toBe(false)
  //   })
  // })

  describe('when the hide/show button is clicked', () => {
    test('changes the password input type to be text', async () => {
      const passwordInput = SignUpCardWrapper.get(
        '[data-testid="password-input"]',
      )
      const passwordConfirmInput = SignUpCardWrapper.get(
        '[data-testid="password-confirm-input"]',
      )

      expect(passwordInput.attributes('type')).toBe('password')
      expect(passwordConfirmInput.attributes('type')).toBe('password')

      await SignUpCardWrapper.get('[data-testid="hide-show-button"]').trigger(
        'click',
      )

      expect(passwordInput.attributes('type')).toBe('text')
      expect(passwordConfirmInput.attributes('type')).toBe('text')
    })
  })

  describe('when the login button is clicked', () => {
    test('emits the log in click event', async () => {
      await SignUpCardWrapper.get('[data-testid="login-button"]').trigger(
        'click',
      )

      expect(SignUpCardWrapper.emitted()).toHaveProperty('logInClick')
    })
  })

  // describe('when the sign up button is clicked', () => {
  //   // const emailAddress = 'strongbad@homestarrunner.com'
  //   // const password = 'homestarsux'
  //   // const username = 'strongbad'

  //   test('emits the sign up click event', async () => {
  //     await SignUpCardWrapper.get('[data-testid="sign-up-button"]').trigger(
  //       'click',
  //     )

  //     expect(SignUpCardWrapper.emitted().signUpClick).toBeTruthy()
  //     // expect(emitted().signUpClick).toBeTruthy()
  //   })

  //   test('clears the state', async () => {
  //     const { getByTestId } = stubbedRender(SignUpCard)
  //     const emailInput: HTMLInputElement = getByTestId('email-input')
  //     const passwordInput: HTMLInputElement = getByTestId('password-input')
  //     const passwordConfirmInput: HTMLInputElement = getByTestId(
  //       'password-confirm-input',
  //     )
  //     const usernameInput: HTMLInputElement = getByTestId('username-input')
  //     await fireEvent.type(emailInput, emailAddress)
  //     await fireEvent.type(usernameInput, username)
  //     await fireEvent.type(passwordInput, password)
  //     await fireEvent.type(passwordConfirmInput, password)
  //     expect(emailInput.value).toBe(emailAddress)
  //     expect(usernameInput.value).toBe(username)
  //     expect(passwordInput.value).toBe(password)
  //     expect(passwordConfirmInput.value).toBe(password)
  //     await fireEvent.click(getByTestId('sign-up-button'))
  //     expect(emailInput.value).toBe('')
  //     expect(usernameInput.value).toBe('')
  //     expect(passwordInput.value).toBe('')
  //     expect(passwordConfirmInput.value).toBe('')
  //   })
  //   test('calls the api', async () => {
  //     const { getByTestId } = stubbedRender(SignUpCard)
  //     const emailInput: HTMLInputElement = getByTestId('email-input')
  //     const passwordInput: HTMLInputElement = getByTestId('password-input')
  //     const passwordConfirmInput: HTMLInputElement = getByTestId(
  //       'password-confirm-input',
  //     )
  //     const usernameInput: HTMLInputElement = getByTestId('username-input')
  //     await fireEvent.type(emailInput, emailAddress)
  //     await fireEvent.type(usernameInput, username)
  //     await fireEvent.type(passwordInput, password)
  //     await fireEvent.type(passwordConfirmInput, password)
  //     await fireEvent.click(getByTestId('sign-up-button'))
  //     const apiCall = fetchMock.mock.calls[0]
  //     expect(apiCall?.[0]).toEqual(
  //       `${process.env.BACKEND_BASE_URL}/api/Users/register`,
  //     )
  //     expect(apiCall?.[1]?.method).toEqual('POST')
  //     expect(apiCall?.[1]?.body).toEqual(
  //       JSON.stringify({
  //         email: emailAddress,
  //         password,
  //         passwordConfirm: password,
  //         username,
  //       }),
  //     )
  //   })
  //   })
})
