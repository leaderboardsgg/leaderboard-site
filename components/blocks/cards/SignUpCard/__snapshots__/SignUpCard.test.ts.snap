// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignUpCard /> renders correctly 1`] = `
<div
  class="card bg-white w-full max-w-xl"
  data-testid="sign-up-card"
  id="signUpCard"
>
  <div
    class="card-header flex flex-row space-x-3"
  >
    <button
      class="base-button 
        flex flex-1
        justify-center
        bg-white
        text-gray-900
        border border-gray-300
      "
      id="login-button"
    >
      
      Log In
    
    </button>
     
    <div
      class="
        flex flex-1
        justify-center
        px-3
        py-2
        rounded
        bg-gray-100
        text-gray-900
      "
    >
      
      Sign Up
    
    </div>
     
    <button
      class="base-button close-button"
      style="display: none;"
    >
      <svg
        class="h-5 w-5"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
   
  <div
    class="p-3"
  >
    <div
      class="flex flex-col space-y-3 pb-3 mb-3 border-b border-gray-300"
    >
      <input
        autocomplete="email"
        class="input"
        name="email"
        placeholder="Email"
        type="text"
      />
       
      <div
        class="flex flex-1 flex-col space-y-1"
      >
        <input
          autocomplete="nickname"
          class="input"
          name="username"
          placeholder="Username"
          type="text"
        />
         
        <p
          class="text-sm"
        >
          This can be changed later
        </p>
      </div>
       
      <div
        class="flex flex-1 flex-col space-y-1"
      >
        <div
          class="
            flex flex-1 flex-col
            sm:flex-row
            space-y-3
            sm:space-y-0 sm:space-x-2
          "
        >
          <input
            autocomplete="password"
            class="input flex-grow"
            name="password"
            placeholder="Password"
            type="password"
          />
           
          <input
            autocomplete="password"
            class="input flex-grow"
            name="confirmPassword"
            placeholder="Confirm"
            type="password"
          />
           
          <button
            class="base-button hidden sm:inline bg-white"
            data-testid="hide-show-password-button"
            id="hide-show-password"
            type="button"
          >
            <svg
              class="h-5 w-5"
              data-testid="hidden-eye-icon"
              fill="currentColor"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M3.707 2.293a1 1 0 00-1.414 1.414l14 14a1 1 0 001.414-1.414l-1.473-1.473A10.014 10.014 0 0019.542 10C18.268 5.943 14.478 3 10 3a9.958 9.958 0 00-4.512 1.074l-1.78-1.781zm4.261 4.26l1.514 1.515a2.003 2.003 0 012.45 2.45l1.514 1.514a4 4 0 00-5.478-5.478z"
                fill-rule="evenodd"
              />
               
              <path
                d="M12.454 16.697L9.75 13.992a4 4 0 01-3.742-3.741L2.335 6.578A9.98 9.98 0 00.458 10c1.274 4.057 5.065 7 9.542 7 .847 0 1.669-.105 2.454-.303z"
              />
            </svg>
             
            <svg
              class="h-5 w-5"
              data-testid="visible-eye-icon"
              fill="currentColor"
              style="display: none;"
              viewBox="0 0 20 20"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M10 12a2 2 0 100-4 2 2 0 000 4z"
              />
               
              <path
                clip-rule="evenodd"
                d="M.458 10C1.732 5.943 5.522 3 10 3s8.268 2.943 9.542 7c-1.274 4.057-5.064 7-9.542 7S1.732 14.057.458 10zM14 10a4 4 0 11-8 0 4 4 0 018 0z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
         
        <p
          class="text-sm"
        >
          
          * Must be 8 characters with a mix of letters and numbers
        
        </p>
      </div>
       
      <button
        class="base-button signup-button"
      >
         Sign Up 
      </button>
    </div>
     
    <div
      class="flex flex-col w-full space-y-2"
    >
      <button
        class="base-button signup-button"
      >
        <svg
          class="w-5 h-5 mr-2"
          viewBox="0 0 480 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M186.1 328.7c0 20.9-10.9 55.1-36.7 55.1s-36.7-34.2-36.7-55.1 10.9-55.1 36.7-55.1 36.7 34.2 36.7 55.1zM480 278.2c0 31.9-3.2 65.7-17.5 95-37.9 76.6-142.1 74.8-216.7 74.8-75.8 0-186.2 2.7-225.6-74.8-14.6-29-20.2-63.1-20.2-95 0-41.9 13.9-81.5 41.5-113.6-5.2-15.8-7.7-32.4-7.7-48.8 0-21.5 4.9-32.3 14.6-51.8 45.3 0 74.3 9 108.8 36 29-6.9 58.8-10 88.7-10 27 0 54.2 2.9 80.4 9.2 34-26.7 63-35.2 107.8-35.2 9.8 19.5 14.6 30.3 14.6 51.8 0 16.4-2.6 32.7-7.7 48.2 27.5 32.4 39 72.3 39 114.2zm-64.3 50.5c0-43.9-26.7-82.6-73.5-82.6-18.9 0-37 3.4-56 6-14.9 2.3-29.8 3.2-45.1 3.2-15.2 0-30.1-.9-45.1-3.2-18.7-2.6-37-6-56-6-46.8 0-73.5 38.7-73.5 82.6 0 87.8 80.4 101.3 150.4 101.3h48.2c70.3 0 150.6-13.4 150.6-101.3zm-82.6-55.1c-25.8 0-36.7 34.2-36.7 55.1s10.9 55.1 36.7 55.1 36.7-34.2 36.7-55.1-10.9-55.1-36.7-55.1z"
            fill="currentColor"
          />
        </svg>
         
        <p>
          Sign Up with Github
        </p>
      </button>
       
      <button
        class="base-button signup-button"
      >
        <svg
          class="w-5 h-5 mr-2"
          viewBox="0 0 488 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M488 261.8C488 403.3 391.1 504 248 504 110.8 504 0 393.2 0 256S110.8 8 248 8c66.8 0 123 24.5 166.3 64.9l-67.5 64.9C258.5 52.6 94.3 116.6 94.3 256c0 86.5 69.1 156.6 153.7 156.6 98.2 0 135-70.4 140.8-106.9H248v-85.3h236.1c2.3 12.7 3.9 24.9 3.9 41.4z"
            fill="currentColor"
          />
        </svg>
         
        <p>
          Sign Up with Google
        </p>
      </button>
    </div>
  </div>
</div>
`;
